crates/lunatic-control-axum/src/routes.rs

```diff
@@ -0,0 +1,154 @@
+        .map_err(|e| ApiError::log_internal("Error generating random token for registration", e))?;
+    let authentication_token = base64_url::encode(&authentication_token);
+
+    control.register(&reg, &cert_pem, &authentication_token);
+
+    ok(Registration {
+        node_name: reg.node_name,
+        cert_pem,
+        authentication_token,
+        root_cert: TEST_ROOT_CERT.into(),
+        urls: ControlUrls {
+            api_base: format!("http://{host}/"),
+            nodes: format!("http://{host}/nodes"),
+            node_started: format!("http://{host}/started"),
+            node_stopped: format!("http://{host}/stopped"),
+            get_module: format!("http://{host}/module/{{id}}"),
+            add_module: format!("http://{host}/module"),
+            get_nodes: format!("http://{host}/nodes"),
+        },
+    })
+}
+
+pub async fn node_stopped(
+    node_auth: NodeAuth,
+    control: Extension<Arc<ControlServer>>,
+) -> ApiResponse<()> {
+    log::info!("Node {} stopped", node_auth.node_name);
+
+    let control = control.as_ref();
+    control.stop_node(node_auth.registration_id as u64);
+
+    ok(())
+}
+
+pub async fn node_started(
+    node_auth: NodeAuth,
+    control: Extension<Arc<ControlServer>>,
+    Json(data): Json<NodeStart>,
+) -> ApiResponse<NodeStarted> {
+    let control = control.as_ref();
+    control.stop_node(node_auth.registration_id as u64);
+
+    let (node_id, _node_address) = control.start_node(node_auth.registration_id as u64, data);
+
+    log::info!("Node {} started with id {}", node_auth.node_name, node_id);
+
+    // TODO spawn all modules on node
+
+    ok(NodeStarted {
+        node_id: node_id as i64,
+    })
+}
+
+pub async fn list_nodes(
+    _node_auth: NodeAuth,
+    control: Extension<Arc<ControlServer>>,
+) -> ApiResponse<NodesList> {
+    let control = control.as_ref();
+    let nds: Vec<_> = control
+        .nodes
+        .iter()
+        .filter(|n| n.status < 2 && !n.node_address.is_empty())
+        .collect();
+    let nodes: Vec<_> = control
+        .registrations
+        .iter()
+        .filter_map(|r| {
+            nds.iter()
+                .find(|n| n.registration_id == *r.key())
+                .map(|n| NodeInfo {
+                    id: *n.key(),
+                    address: n.node_address.parse().unwrap(),
+                    name: r.node_name.to_string(),
+                })
+        })
+        .collect();
+
+    ok(NodesList { nodes })
+}
+
+pub async fn add_module(
+    node_auth: NodeAuth,
+    control: Extension<Arc<ControlServer>>,
+    body: Bytes,
+) -> ApiResponse<ModuleId> {
+    log::info!("Node {} add_module", node_auth.node_name);
+
+    let control = control.as_ref();
+    let module_id = control.add_module(body.to_vec());
+    ok(ModuleId { module_id })
+}
+
+pub async fn get_module(
+    node_auth: NodeAuth,
+    PathExtractor(id): PathExtractor<u64>,
+    control: Extension<Arc<ControlServer>>,
+) -> ApiResponse<ModuleBytes> {
```

Comment:
This returns the module as json with `Vec<u8>`, we should probably return something like `body::Bytes`

Reply:
I agree but let's leave it for now, we can change it later.